File: backend/lib/User/UserRepository.php
<?php

namespace lib\User;

use lib\Core\Repository;

class UserRepository extends Repository
{
    public function __construct()
    {
        parent::__construct();
        $this->createTable();
    }

    private function createTable(): void
    {
        $this->db->exec("CREATE TABLE IF NOT EXISTS users (
            id            TEXT    PRIMARY KEY,
            email         TEXT    NOT NULL UNIQUE,
            password      TEXT    NOT NULL,
            name          TEXT    NOT NULL DEFAULT '',
            avatarUrl     TEXT    NOT NULL DEFAULT '',
            profileImages TEXT    NOT NULL DEFAULT '[]',
            createdAt     INTEGER NOT NULL
        )");
        try { $this->db->exec("ALTER TABLE users ADD COLUMN avatarUrl TEXT NOT NULL DEFAULT ''"); } catch (\Throwable) {}
        try { $this->db->exec("ALTER TABLE users ADD COLUMN profileImages TEXT NOT NULL DEFAULT '[]'"); } catch (\Throwable) {}
    }

    public function create(UserEntity $user): bool
    {
        return (bool) $this->execute(
            "INSERT INTO users (id, email, password, name, avatarUrl, profileImages, createdAt)
             VALUES (?, ?, ?, ?, ?, ?, ?)",
            [$user->id, $user->email, $user->password, $user->name,
             $user->avatarUrl, json_encode($user->profileImages), $user->createdAt]
        );
    }

    public function findById(string $id): ?array
    {
        return $this->fetch("SELECT * FROM users WHERE id = ?", [$id]);
    }

    public function findByEmail(string $email): ?array
    {
        return $this->fetch("SELECT * FROM users WHERE email = ?", [$email]);
    }

    public function existsById(string $id): bool
    {
        return (bool) $this->fetch("SELECT id FROM users WHERE id = ?", [$id]);
    }

    public function existsByEmail(string $email): bool
    {
        return (bool) $this->fetch("SELECT id FROM users WHERE email = ?", [$email]);
    }

    public function updateById(UserEntity $user): bool
    {
        return (bool) $this->execute(
            "UPDATE users SET name = ?, avatarUrl = ?, profileImages = ? WHERE id = ?",
            [$user->name, $user->avatarUrl, json_encode($user->profileImages), $user->id]
        );
    }

    public function deleteById(string $id): bool
    {
        return (bool) $this->execute("DELETE FROM users WHERE id = ?", [$id]);
    }
}

File: frontend/lib/repositories/user/user.repository.dart
import 'dart:convert';
import 'package:frontend/api/api.client.dart';

class UserRepository {
  static final instance = UserRepository._();
  UserRepository._();

  Future<Map<String, dynamic>> register({
    required String email,
    required String password,
    String? name,
  }) =>
      ApiClient.instance.post('/api.php', {
        'method': 'user.register',
        'email': email,
        'password': password,
        if (name != null) 'name': name,
      });

  Future<Map<String, dynamic>> login({
    required String email,
    required String password,
  }) =>
      ApiClient.instance.post('/api.php', {
        'method': 'user.login',
        'email': email,
        'password': password,
      });

  Future<Map<String, dynamic>> getById(String id) =>
      ApiClient.instance.post('/api.php', {'method': 'user.findById', 'id': id});

  Future<Map<String, dynamic>> updateById({
    required String id,
    String? name,
    List<Map<String, dynamic>>? profileImages,
  }) =>
      ApiClient.instance.post('/api.php', {
        'method': 'user.updateById',
        'id': id,
        if (name != null) 'name': name,
        if (profileImages != null) 'profileImages': jsonEncode(profileImages),
      });

  Future<Map<String, dynamic>> deleteById(String id) =>
      ApiClient.instance.post('/api.php', {'method': 'user.deleteById', 'id': id});
}
